src/hotspot/share/gc/shenandoah/shenandoahBarrierSet.inline.hpp

```diff
@@ -154,9 +154,12 @@ inline void ShenandoahBarrierSet::enqueue(oop obj) {
 template <DecoratorSet decorators, typename T>
 inline void ShenandoahBarrierSet::satb_barrier(T *field) {
   if (HasDecorator<decorators, IS_DEST_UNINITIALIZED>::value ||
-      HasDecorator<decorators, AS_NO_KEEPALIVE>::value) {
+      HasDecorator<decorators, AS_NO_KEEPALIVE>::value ||
+      HasDecorator<decorators, ON_WEAK_OOP_REF>::value ||
+      HasDecorator<decorators, ON_PHANTOM_OOP_REF>::value) {
```

Comment:
Suggest to split it into two things, with comments:

```
  // Uninitialized and no-keepalive stores do not need barrier.
  if (HasDecorator<decorators, IS_DEST_UNINITIALIZED>::value ||
      HasDecorator<decorators, AS_NO_KEEPALIVE>::value) {
    return;
  }

  // Stores to weak/phantom require no barrier. The old references would
  // have been resurrected by load barrier if they were needed.
  if (HasDecorator<decorators, ON_WEAK_OOP_REF>::value ||
      HasDecorator<decorators, ON_PHANTOM_OOP_REF>::value) {
    return;
  }
```

(I think I caught the reason why we are safe to skip SATB here, maybe comment can be expanded)

Reply:
Ha. I had it that way originally - I'll put it back.